### PR TITLE
An Alternative to "Use latest Azure Storage libraries"

### PR DESCRIPTION
**Pull request by martins-vds**
Dependent on #35 

This PR is inspired by #33 and proposes a slightly different implementation.

### CODE REVIEW

**Review by stebet — COMMENTED** [review]
Same review comments as for #35

**Comment by martins-vds** [comment]
Just integrated changes made for #35

File: Functions/Program.cs [hunk 1]
```diff
@@ -10,7 +11,21 @@ public static void Main()
...
+                    var storageConnectionString = context.Configuration["PwnedPasswordsConnectionString"];
+                    var storageContainerName = context.Configuration["BlobContainerName"];
+
+                    services.Configure<BlobStorageOptions>(options => options.BlobContainerName = storageContainerName);
+
+                    services.AddAzureClients(
+                        azure =>
+                        {
+                            azure.AddBlobServiceClient(storageConnectionString);
+                        });
```
Comment (stebet) [inline Functions/Program.cs:25]:
```suggestion
                    services.AddAzureClients(azure => azure.AddBlobServiceClient(storageConnectionString));
```

File: Functions/BlobStorageOptions.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Functions
+{
+    public class BlobStorageOptions
+    {
+        public string BlobContainerName { get; set; }
+    }
```
Comment (stebet) [inline Functions/BlobStorageOptions.cs:6]:
```suggestion
    public class BlobStorageOptions : IOptions<BlobStorageOptions>
    {
        public string BlobContainerName { get; set; }
        public BlobStorageOptions Value => this;
    }
```

File: Functions.Tests/Helpers/TestFunctionDefinition.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            {
+                outputs.Add($"outputName{i}", new TestBindingMetadata($"TestOutput{i}", BindingDirection.Out));
+            }
+
+            for (int i = 0; i < paramTypes.Length; i++)
+            {
+                var properties = new Dictionary<string, object>
+                {
+                    {"TestPropertyKey", "TestPropertyValue" }
+                };
```
Comment (stebet) [inline Functions.Tests/Helpers/TestFunctionDefinition.cs:73]: What's the point of this dictionary?

**Review by stebet — CHANGES_REQUESTED** [review]

File: Functions.Tests/Helpers/TestFunctionDefinition.cs (hunk 3 above)
Comment (martins-vds) [reply Functions.Tests/Helpers/TestFunctionDefinition.cs]: I made it static

File: .github/dependabot.yml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+version: 2
+updates:
+  - package-ecosystem: "nuget"
+    directory: "/"
+    schedule:
+      interval: "daily"
+      time: "10:00"
+      timezone: "Australia/Brisbane"
```
Comment (stebet) [inline .github/dependabot.yml:8]: This change shouldn't be showing up in the PR as it's already in `main`

File: .github/workflows/ci-build.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      with:
+        name: CoverageReport # Artifact name        
+        path: coveragereport # Directory containing files to upload
+    
+    - name: Upload Test Results
+      uses: actions/upload-artifact@v2  # upload test results
+      if: always()                      # always run this step
+      with:
+        name: test-results
+        path: Functions.Tests/TestResults/test-results.trx
```
Comment (stebet) [inline .github/workflows/ci-build.yml:60]: This change shouldn't be showing up in the PR as it's already in `main`

File: .github/workflows/publish-test-results.yml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+jobs:
+  report:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: dorny/test-reporter@v1
+      with:
+        artifact: test-results            # artifact name
+        name: dotnet test results         # Name of the check run which will be created
+        path: '*.trx'                     # Path to test results (inside artifact .zip)
+        reporter: dotnet-trx              # Format of test results
```
Comment (stebet) [inline .github/workflows/publish-test-results.yml:16]: This change shouldn't be showing up in the PR as it's already in `main`

File: Functions/BlobStorage.cs [hunk 7]
```diff
@@ -1,65 +1,67 @@
...
+                _log.LogInformation("Hash file downloaded in {ElapsedMilliseconds}ms", sw.ElapsedMilliseconds.ToString("n0"));
+
+                return new BlobStorageEntry(response.Value.Content, response.Value.Details.LastModified);
             }
-            catch (StorageException ex) when (ex.RequestInformation?.HttpStatusCode == 404)
+            catch (RequestFailedException ex) when (ex.Status == 404)
             {
-                _log.LogWarning("Blob Storage couldn't find file \"{FileName}\"", fileName);
+                _log.LogWarning("Hash file \"{FileName}\" not found.", fileName);
             }
```
Comment (stebet) [inline Functions/BlobStorage.cs:65]: These changes seem out of date with what is already on `main`

File: Functions/Functions.csproj [hunk 8]
```diff
@@ -5,11 +5,12 @@
     <OutputType>Exe</OutputType>
   </PropertyGroup>
   <ItemGroup>
+    <PackageReference Include="Azure.Storage.Blobs" Version="12.9.0" />
     <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http" Version="3.0.13" />
     <PackageReference Include="Microsoft.Azure.Functions.Worker.Sdk" Version="1.0.3" OutputItemType="Analyzer" />
-    <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.2.0" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.3.0" />
+    <PackageReference Include="Microsoft.Extensions.Azure" Version="1.0.0" />
```
Comment (stebet) [inline Functions/Functions.csproj:12]: This seems out of date from what is already on `main`.

**Review by stebet — CHANGES_REQUESTED** [review]
A lot of code seems to be out of date from `main`, almost like a rebase went wrong. It might be easier to create a new PR from a fresh `main` fork with the changes instead? I think another rebase + force push might get messy.

**Comment by martins-vds** [comment]
@stebet I'll close this one, wait for #42 to get merged and then I'll resubmit this PR

**Pull request closed without merging**